### PR TITLE
fix(ivy): wrong context passed to ngOnDestroy when resolved multiple times

### LINKED ISSUE #35167: Wrong 'this' in a service using Ivy
# 🐞 bug report

### Affected Package

 @angular/core

### Is this a regression?

 Yes, the previous version in which this bug was not present was: 8

### Description

 I have a service abstraction over ngOnDestroy to unsubscribe from streams. Here's its source code, it's pretty simple:
```
@Injectable()
export class DestroyService extends Observable<void> implements OnDestroy {
    private readonly life$ = new Subject<void>();

    constructor() {
        super(subscriber => this.life$.subscribe(subscriber));
    }

    ngOnDestroy() {
        this.life$.next();
        this.life$.complete();
    }
}
```
I have it in my library of UI components that I compile with Angular 7/8. After I tried it with Angular 9 RC application (tried with 14 and with some old ones) — when a component (from my library) using that service gets destroyed I get an error `Cannot read property 'next' of undefined`. Setting breakpoints shown me that `this` inside this service is, for some reason, component instance that uses it, not the service instance.

## 🔬 Minimal Reproduction

 I was unable to reproduce it in an isolated environment but I have no clue how this could have happened. My code is proprietary which I'm struggling to open source but it takes a while. I was hoping maybe somebody from Angular team who knows Ivy compiler and ngcc have any ideas how this could have theoretically happened.

## 🔥 Exception or Error
This inside ngOnDestroy of a service is component instance, not service instance

## 🌍  Your Environment

**Angular Version:**
<pre><code>Angular 9.0.0-rc.14!
</code></pre>

**Anything else relevant?**
Sorry for such a vague description, the library is pretty straightforward Angular CLI library compiled with ng-packagr. It's just really huge and when I try to make a small library with relevant code — the issue does not reproduce. Without deep knowledge of how ngOnDestroy gets called I failed to debug it on my side.

### LINKED ISSUE #35231: Invalid ngOnDestroy invocation on multiple provision
# 🐞 bug report
### Affected Package
core? 9.0.0. Would like to know
### Is this a regression?
It occured after 

 Yes, the previous version in which this bug was not present was: ....

### Description
Not sure whether it is connected with https://github.com/angular/angular/issues/35167 or not, but following screen shows an error:
image
in summary, those services in array - all of them implements some interface and are provided with the same token, under provision with `multiple: true`. As you see, `ngOnDestroy` is invoked on that array, but this `this`... eh, I don't think that this array extends `Subscriber`. I have also noticed that this is not always the case.

## 🔥 Exception or Error
<pre><code>
ERROR Error: Uncaught (in promise): TypeError: this.clearSubscriptions is not a function
TypeError: this.clearSubscriptions is not a function
    at Array.ngOnDestroy (subscriber.ts:28)
    at executeOnDestroys (core.js:14024)
    at cleanUpView (core.js:13931)
    at destroyViewTree (core.js:13705)
    at destroyLView (core.js:13882)
    at RootViewRef.destroy (core.js:14737)
    at ComponentRef$1.destroy (core.js:33528)
    at RouterOutlet.deactivate (router.js:8912)
    at ActivateRoutes.deactivateRouteAndOutlet (router.js:3963)
    at ActivateRoutes.deactivateRouteAndItsChildren (router.js:3921)
    at resolvePromise (zone-evergreen.js:793)
    at resolvePromise (zone-evergreen.js:752)
    at zone-evergreen.js:854
    at ZoneDelegate.invokeTask (zone-evergreen.js:400)
    at Object.onInvokeTask (core.js:40744)
    at ZoneDelegate.invokeTask (zone-evergreen.js:399)
    at Zone.runTask (zone-evergreen.js:168)
    at drainMicroTaskQueue (zone-evergreen.js:570)
    at invokeTask (zone-evergreen.js:485)
    at ZoneTask.invoke (zone-evergreen.js:470)
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.1
Node: 12.13.1
OS: win32 x64

Angular: 9.0.0
... animations, cdk, common, compiler, compiler-cli, core
... elements, forms, language-service, material
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                            Version
------------------------------------------------------------
@angular-devkit/architect          0.900.1
@angular-devkit/build-angular      0.900.1
@angular-devkit/build-ng-packagr   0.900.1
@angular-devkit/build-optimizer    0.900.1
@angular-devkit/build-webpack      0.900.1
@angular-devkit/core               8.3.23
@angular-devkit/schematics         8.3.23
@angular/cli                       9.0.1
@ngtools/webpack                   9.0.1
@schematics/angular                8.3.23
@schematics/update                 0.900.1
ng-packagr                         9.0.0
rxjs                               6.5.4
typescript                         3.7.5
webpack                            4.41.2
</code></pre>

### PR DESCRIPTION
**Pull request by crisbeto**
When the same provider is resolved multiple times on the same node, the first invocation had the correct context, but all subsequent ones were incorrect because we were registering the hook multiple times under different indexes in `destroyHooks`.

Fixes #35167.

### CODE REVIEW

File: packages/core/src/render3/di_setup.ts [hunk 1]
```diff
@@ -81,26 +86,25 @@ function resolveProvider(
...
           endIndex);
-      if (existingFactoryIndex == -1) {
+      if (existingFactoryIndex === -1) {
         diPublicInInjector(
             getOrCreateNodeInjectorForNode(
                 tNode as TElementNode | TContainerNode | TElementContainerNode, lView),
             tView, token);
+        if (providerIsTypeProvider || isClassProvider(provider)) {
+          const prototype = ((provider as ClassProvider).useClass || provider).prototype;
+          const ngOnDestroy = prototype.ngOnDestroy;
```
Comment (Azbesciak) [inline packages/core/src/render3/di_setup.ts]: What with `useFactory` for example when you provide some object which is not a class instance?

**Comment by Azbesciak** [comment]
@crisbeto yeah, you are right. That would also explain why I always call `ngOnDestroy` from my root components manually also... because I noticed that this is sometimes called, sometimes not.

File: packages/core/src/render3/di_setup.ts [hunk 2]
```diff
@@ -17,10 +17,12 @@ import {DirectiveDef} from './interfaces/definition';
 import {NodeInjectorFactory} from './interfaces/injector';
 import {TContainerNode, TDirectiveHostNode, TElementContainerNode, TElementNode, TNodeProviderIndexes} from './interfaces/node';
 import {isComponentDef} from './interfaces/type_checks';
-import {LView, TData, TVIEW, TView} from './interfaces/view';
+import {LView, TData, TVIEW} from './interfaces/view';
 import {getLView, getPreviousOrParentTNode, getTView} from './state';
 
 
+/** Any provider definition that's not a `TypeProvider`. */
+type NonTypeProvider = Exclude<Exclude<Provider, TypeProvider>, any[]>;
```
Comment (waterplea) [inline packages/core/src/render3/di_setup.ts]: Wouldn't it work with a single Exclude?
`Exclude<Provider, TypeProvider | any[]>`

File: packages/core/src/render3/di_setup.ts (hunk 1 above)
Comment (crisbeto) [reply packages/core/src/render3/di_setup.ts]: I'm not sure about that one, but I'll bring it up. The previous approach didn't support `useFactory` either.

**Review by MarsiBarsi — APPROVED** [review]

**Review by waterplea — APPROVED** [review]

**Comment by johannesnormannjensen** [comment]
This fix is really needed

**Comment by Azbesciak** [comment]
@pkozlowski-opensource could you... :)

**Review by kara — APPROVED** [review]
LGTM

**Comment by kara** [comment]
[presubmit](http://test/OCL:295839203:BASE:295839229:1582069177109:8308dc40)

**Pull request closed without merging**